tensorrt_llm/_torch/pyexecutor/llm_request.py

```diff
@@ -35,10 +35,199 @@
+        if self.beam_width == -1:
+            self._init(new_probs)
+
+        assert len(new_probs) == self.beam_width, "Beam width mismatch"
+        for idx, probs in enumerate(new_probs):
+            self.log_probs[idx].extend(probs)
+            self.cum_log_probs[idx] += sum(probs)
+
+
+class PyResult:
+    """PyResult reimplements some features of `bindings.executor.Result` in Python"""
+
+    def __init__(self,
+                 prompt_len: int,
+                 max_new_tokens: int,
+                 device_memory=False,
+                 streaming=False,
+                 return_log_probs: bool = False,
+                 return_context_logits: bool = False,
+                 return_generation_logits: bool = False):
+        self._streaming = streaming
+        self._context_logits = LogitsStorage(
+            prompt_len, device_memory) if return_context_logits else None
+        self._generation_logits = LogitsStorage(
+            max_new_tokens, device_memory) if return_generation_logits else None
+        self._log_probs = LogProbStorage() if return_log_probs else None
+
+    def append_context_logits(self, context_logits: torch.Tensor):
+        if self._context_logits:
+            self._context_logits.append(context_logits)
+
+    def append_generation_logits(self, generation_logits: torch.Tensor):
+        if self._generation_logits:
+            self._generation_logits.append(generation_logits)
+
+    def append_log_probs(self, log_probs: LogProbs):
+        if self._log_probs:
+            self._log_probs.append(log_probs)
+
+    @property
+    def context_logits(self) -> torch.Tensor | None:
+        return self._context_logits and self._context_logits.get(
+            True)[:, 0]  # remove beam_width axis for context
+
+    @property
+    def generation_logits(self) -> torch.Tensor | None:
+        # Internal storage: [seq_length, beam_width, vocab_size]
+        # API expect: [beam_width, seq_length, vocab_size]
+        return self._generation_logits and self._generation_logits.get(
+            not self._streaming).transpose(0, 1)
+
+    @property
+    def log_probs(self) -> list[list[float]] | None:
+        return self._log_probs and self._log_probs.log_probs
+
+    @property
+    def cum_log_probs(self) -> list[float] | None:
+        return self._log_probs and self._log_probs.cum_log_probs
+
+
+class LlmResult:
+    """LlmResult wraps `bindings.executor.Result` but detour some features to Python implementation"""
+    py_result_properties = frozenset(
+        ('context_logits', 'generation_logits', 'log_probs', 'cum_log_probs'))
+
+    def __init__(self, result: tensorrt_llm.bindings.executor.Result,
+                 py_result: PyResult):
+        self._result = result
+        self._py_result = py_result
+
+    def __getattr__(self, item):
+        if item in self.py_result_properties:
+            return getattr(self._py_result, item)
+        return getattr(self._result, item)
+
+
+class LlmResponse:
```

Comment:
Maybe a minor issue, it is a bit odd to have `LlmResponse` in a file called `llm_request.py`, I thought it is for the `LlmRequest`. Should we partition the file or make the file name proper for a package of PyExecutor's input and output DS?

Reply:
I don't have preference on this. @QiJune any opinion?

Reply:
Yes, I think we can move it to `llm_response.py`. Let's do it in the following PR.